reana_workflow_engine_snakemake/utils.py

```diff
@@ -17,7 +17,7 @@ def publish_workflow_start(
     workflow_uuid: str, publisher: WorkflowStatusPublisher, job: Job
 ):
     """Publish to MQ the start of the workflow."""
-    job_count = len([rule for rule in job.dag.rules if not rule.norun])
+    job_count = len([j for j in job.dag.needrun_jobs if not j.rule.norun])
```

Comment:
I was looking at the source code of snakemake ([source](https://github.com/snakemake/snakemake/blob/v6.8.0/snakemake/dag.py#L380-L383))
```python
@property
def needrun_jobs(self):
    """Jobs that need to be executed."""
    return filterfalse(self.finished, self._needrun)
```

It seems like `needrun_jobs` returns the job that need to be run and that haven't finished yet. Maybe we should use `_needrun` directly here, even though it's internal?

Otherwise what's happening is that we are sending the total multiple times, often with a different total number because some jobs have finished in the meantime. I have to say that it works even as it is now, but just because we have this check in r-w-controller ([source](https://github.com/reanahub/reana-workflow-controller/blob/d48ef4665a3f4a358ec0958a140f3cc6794ffd85/reana_workflow_controller/consumer.py#L206-L212))
```python
            if previous_status:
                previous_total = previous_status.get("total") or 0
            if status == "total":
                if previous_total > 0:
                    continue
                else:
                    job_progress["total"] = msg["progress"]["total"]

```
The first message brings the right total amount, so that is saved to the database and is never updated (at least, that's what I think it's happening).

Reply:
Nice catch! It indeed appears that only the first received total is considered by REANA, probably for this exact reason.

I investigated a bit more and, if we choose to use the internal variables and not rely on this check from r-w-controller, it's probably better to use `dag._finished | dag._needrun` instead of just `dag._needrun` (as done by Snakemake to [check the length of the DAG](https://github.com/snakemake/snakemake/blob/26373e66fbaba04fa076aebb24ad869f41f228b9/snakemake/dag.py#L1053)). This is because sometimes the `update_needrun` method might be called (e.g. for dynamic workflows or whenever there is a change to the DAG topology) and the jobs that are already finished are removed from the `_needrun` set.